File: agent-driven-architecture-main/examples/typescript/src/blocks/inbox/project.ts
// ── blocks/inbox/project — TWO projections, and the second one is the point ─
// The view row is for the operator. The CONTEXT LINE is for the model, and it is
// what makes "observable, never silent" mean something: the reasoner is told, in
// its own input, that the system is shedding load on its behalf.
//
//     "2 input(s) conflated from ticket-stream"
//     "1 turn(s) failed: backend timeout"
//
// An agent that does not know it dropped two inputs will confidently answer as
// if it saw them.

import { bounded, MAX_CONTEXT_LINES_PER_BLOCK } from "@adr/spine/pure/context";
import type { InboxSlice } from "./slice";

export interface InboxRow {
  readonly source: string;
  readonly conflated: number;
  readonly duplicates: number;
}

export interface InboxView {
  readonly rows: readonly InboxRow[];
  readonly faults: readonly string[];
}

export function inboxView(slice: InboxSlice): InboxView {
  const sources = [...new Set([...slice.conflated.keys(), ...slice.duplicates.keys()])];
  return {
    rows: sources.map((source) => ({
      source,
      conflated: slice.conflated.get(source) ?? 0,
      duplicates: slice.duplicates.get(source) ?? 0,
    })),
    faults: slice.faults.map((f) => `${f.source}: ${f.fault}`),
  };
}

export function inboxContextLines(
  slice: InboxSlice,
  max: number = MAX_CONTEXT_LINES_PER_BLOCK,
): readonly string[] {
  const drops = [...slice.conflated.entries()].map(
    ([source, dropped]) => `${dropped} input(s) conflated from ${source} — you did not see them`,
  );
  const dupes = [...slice.duplicates.entries()].map(
    ([source, count]) => `${count} duplicate input(s) refused from ${source}`,
  );
  const faults = slice.faults.map((f) => `turn failed on ${f.source}: ${f.fault}`);
  return bounded([...drops, ...dupes, ...faults], max);
}

File: agent-driven-architecture-main/examples/kotlin/block/inbox/src/main/kotlin/adr/blocks/inbox/Project.kt
// ── blocks/inbox/project — TWO pure projections of the SAME slice ──────────
// slice → view    what a human reads      (6.9)
// slice → lines   what the reasoner reads (G15)
//
// THE SECOND ONE IS WHAT MAKES "OBSERVABLE, NEVER SILENT" MEAN SOMETHING. A counter
// only an operator can see tells the agent nothing; a context line tells the
// REASONER it is shedding load, so it can decide to be terser, batch, or ask for
// help. 12.2's requirement is not "log the drop", it is "the drop is visible where
// decisions are made" — and both places where decisions are made get told.
//
// Bounded by the same constant every other block uses, so a storm of drops cannot
// inflate the prompt.

package adr.blocks.inbox

import adr.spine.pure.MAX_CONTEXT_LINES_PER_BLOCK

data class InboxView(
    val conflated: List<String>,
    val duplicates: List<String>,
    val faults: List<String>,
)

internal class InboxProjection {

    fun view(slice: InboxSlice): InboxView = InboxView(
        conflated = slice.conflated.map { (source, count) -> "${source.value}: $count conflated" },
        duplicates = slice.duplicates.map { (source, count) -> "${source.value}: $count duplicate" },
        faults = slice.faults,
    )

    fun contextLines(slice: InboxSlice, maxLines: Int = MAX_CONTEXT_LINES_PER_BLOCK): List<String> = (
        slice.conflated.map { (source, count) -> "$count input(s) conflated from ${source.value}" } +
            slice.duplicates.map { (source, count) -> "$count duplicate input(s) from ${source.value}" } +
            slice.faults.map { "turn failed: $it" }
        ).takeLast(maxLines)
}
